system/lib/pthread/proxying.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2021 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+#pragma once
+
+#include <emscripten/emscripten.h>
+#include <pthread.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Opaque handle to a set of thread-local work queues to which work can be
+// asynchronously or synchronously proxied from other threads. When work is
+// proxied to a queue on a particular thread, that thread is notified to start
+// processing work from that queue if it is not already doing so.
+typedef struct em_proxying_queue em_proxying_queue;
+
+// Create and destroy proxying queues.
+em_proxying_queue* em_proxying_queue_create();
+void em_proxying_queue_destroy(em_proxying_queue* q);
+
+// Get the queue used for proxying low-level runtime work. Work on this queue
+// may be processed at any time inside system functions, so it must be
+// nonblocking and safe to run at any time, similar to a native signal handler.
+em_proxying_queue* emscripten_proxy_get_system_queue();
```

Comment:
I interpreted this PR first to offer a user level mechanism for proxying queues? But this function reads like it is intending to supplant the original proxying queue code?

Reply:
In addition to offering user-level queues, the intention is to reimplement the existing API to use this new queue implementation in a follow-up PR, mostly just to reduce the number of queue implementations we have.